docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,295 @@
+    parent    types.KVStore
+    listeners []types.WriteListener
+}
+
+// NewStore returns a reference to a new traceKVStore given a parent
+// KVStore implementation and a buffered writer.
+func NewStore(parent types.KVStore, listeners []types.WriteListener) *Store {
+	return &Store{parent: parent, listeners: listeners}
+}
+
+// Set implements the KVStore interface. It traces a write operation and
+// delegates the Set call to the parent KVStore.
+func (tkv *Store) Set(key []byte, value []byte) {
+	types.AssertValidKey(key)
+	onWrite(tkv.listeners, key, value)
+	tkv.parent.Set(key, value)
+}
+
+// Delete implements the KVStore interface. It traces a write operation and
+// delegates the Delete call to the parent KVStore.
+func (tkv *Store) Delete(key []byte) {
+	onWrite(tkv.listeners, key, nil)
+	tkv.parent.Delete(key)
+}
+
+// onWrite writes a KVStore operation to all of the WriteListeners
+func onWrite(listeners []types.WriteListener, key, value []byte) {
+	for _, l := range listeners {
+		l.OnWrite(key, value)
+	}
+}
+```
+
+### MultiStore interface updates
+We will update the `MultiStore` interface to allow us to wrap a set of listeners around a specific `KVStore`.
+Additionally, we will update the `CacheWrap` and `CacheWrapper` interfaces to enable listening in the caching layer, and add a `MultiStore` method
+to turn on or off this cache listening.
+
+```go
+type MultiStore interface {
+	...
+
+	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
+	ListeningEnabled(key StoreKey) bool
+
+	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
+	SetListeners(key StoreKey, listeners []WriteListener)
+
+	// CacheListening enables or disables KVStore listening at the cache layer
+	CacheListening(listen bool)
+}
+```
+
+```go
+type CacheWrap interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
+}
+
+type CacheWrapper interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
+}
+```
+
+### MultiStore implementation updates
+We will modify all of the `Store` and `MultiStore` implementations to satisfy these new interfaces, and adjust the `rootmulti` `GetKVStore` method
+to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on.
+
+```go
+func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
+	store := rs.stores[key].(types.KVStore)
+
+	if rs.TracingEnabled() {
+		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
+	}
+	if rs.ListeningEnabled(key) {
+		store = listenkv.NewStore(store, rs.listeners[key])
+	}
+
+	return store
+}
+```
+
+We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to enable listening
+in the cache layer when `CacheListening` is turned on.
```

Comment:
Why do we need an explicit `CacheListening`? The root-multistore is always cache-wrapped during block execution. Shouldn't we always do this? Or am I missing something?

Reply:
No I think I'm missing something. The idea was that- assuming some Puts/Deletes in cache layer may never be written to a base Store- we would want to be able to turn on/off listening in the cache layer independently, so that we can choose to only listen to changes that are persisted, only changes in the cache layer, or both.

When the rootmulti is cache-wrapped [here](https://github.com/i-norden/cosmos-sdk/blob/multistore_state_listening/store/rootmulti/store.go#L451) the Stores it forwards are not yet listener-wrapped. Only if enabled will the cachemulti constructor [wrap the Stores in the cache layer with the listeners](https://github.com/i-norden/cosmos-sdk/blob/multistore_state_listening/store/cachemulti/store.go#L60).

Reply:
I think I see my mistake now (one of them, at least). The Stores should be listener wrapped before passing them to the cachemulti constructor. Otherwise there is no listening when cachemulti writes out the Put/Delete operations to the underlying Stores, instead it only listens to the Put/Deletes in the cache layer and only if`CacheListening` is turned on. I still wonder if it would be useful to enable independent listening in the cache and base layer?